### PR TITLE
chore(demand_style_e2e): 1. add test id to constructor

### PR DESCRIPTION
**Pull request by bar-unity**
2. each component should extend base component
                         3. base component should extend base element

### CODE REVIEW

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 import {getTestId, getTestIdSelector} from '../../global/utils';
+import {BaseComponent} from '../base-component';
 
-export class AlertComponent {
+export class AlertComponent extends BaseComponent {
     readonly page: Page;
 
-    constructor(page: Page) {
+    constructor(page: Page, selector: string) {
+        super(page, selector);
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:10]: You are still using playwright methods directly instead of using the wrapped methods under base element page. The whole point of this class is that you will not usr page.click / waitForSelector directly

**Review by NoaMartan — CHANGES_REQUESTED** [review]
Please go over the notes and reach out if something is unclear :)

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 2]
```diff
@@ -1,11 +1,13 @@
 import {Page} from '@playwright/test';
 import {AlertTestIdModifiers} from '@ironsource/fusion-ui/entities';
 import {getTestId, getTestIdSelector} from '../../global/utils';
+import {BaseComponent} from '../base-component';
 
-export class AlertComponent {
+export class AlertComponent extends BaseComponent {
```
[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts:6]: Which methods?

File: projects/fusion-e2e/tests/components/alert/alert-page.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ export class AlertPage extends ComponentBasePage {
         };
 
         super(alertProps);
-        this.component = new AlertComponent(page);
+        this.component = new AlertComponent(page, this.testId);
```
[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-page.ts:19]: I'm not sure I understand. This isn't what you asked me to do?

[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-page.ts:19]: Already done
export class AlertPage extends ComponentBasePage {

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 1 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts:10]: Done

File: projects/fusion-e2e/tests/components/alert/alert-page.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-page.ts:19]: After sync: use test id in super selector

[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-page.ts:19]: after sync: change ComponentBasePage to BasePage

File: projects/fusion-e2e/tests/components/alert/alert-page.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class AlertPage extends ComponentBasePage {
...
     }
 
     getAlertText() {
```
[the remark this answers is not in the document]
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-page.ts:22]: after sync: also in page dont use PW methods

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 5]
```diff
@@ -1,45 +1,59 @@
...
-    constructor(page: Page) {
-        this.page = page;
+import {BaseComponent} from '../base-component';
+import {BaseElement} from '../../behavior';
+import {ButtonComponent} from '../button/button-component';
+
+export class AlertComponent extends BaseComponent {
+    private wrapperElement: BaseElement;
+    private messageElement: BaseElement;
+    private titleElement: BaseElement;
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:10]: All of the component class properties should inherit from a designated class that is not base element. For example title element should implement from StaticText class. Same note applies to wrapperElement and messageElement

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 6]
```diff
@@ -1,45 +1,59 @@
...
-    readonly page: Page;
-
-    constructor(page: Page) {
-        this.page = page;
+import {BaseComponent} from '../base-component';
+import {BaseElement} from '../../behavior';
+import {ButtonComponent} from '../button/button-component';
+
+export class AlertComponent extends BaseComponent {
+    private wrapperElement: BaseElement;
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:8]: Just a naming suggestion, but you don’t have to state 'Element' or 'Component' at the end of everything. For example actionButtonElement can be just actionButton and AlertComponent can be Alert

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 7]
```diff
@@ -1,45 +1,59 @@
...
+            page,
+            getTestIdSelector(getTestId(this.selector, AlertTestIdModifiers.ACTION_BUTTON))
+        );
+        this.closeButtonElement = new ButtonComponent(page, getTestIdSelector(getTestId(this.selector, AlertTestIdModifiers.CLOSE_BUTTON)));
     }
 
     async waitForComponent({testId}: {testId: string}) {
         const loadedPageSelector = getTestIdSelector(getTestId(testId, AlertTestIdModifiers.WRAPPER));
-        await this.page.waitForSelector(loadedPageSelector);
+        await this.waitForSelector(loadedPageSelector);
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:28]: Why does wait for component is implemented here? sounds like a method that should be under baseComponent, also why does it requires a testId param?

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 8]
```diff
@@ -1,45 +1,59 @@
...
     }
 
-    getAlertText({testId}: {testId: string}) {
-        return this.page.getByTestId(getTestId(testId, AlertTestIdModifiers.MESSAGE)).textContent();
+    async getAlertText() {
+        return this.messageElement.textContent();
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:32]: i get this error: Property 'textContent' does not exist on type 'ButtonComponent'.ts(2339)
I checked your buttonComponent class and it seems that the implementation there can use a bit modification.

- class Button should extend from base element (in order for the textContext method to be inherited)
-  class Button's properties should be button: Clickable, label: StaticText;

File: projects/fusion-e2e/tests/components/alert/alert-component.ts [hunk 9]
```diff
@@ -1,45 +1,59 @@
...
     }
 
-    getAlertTitle({testId}: {testId: string}) {
-        return this.page.getByTestId(getTestId(testId, AlertTestIdModifiers.TITLE)).textContent();
+    async getAlertTitle() {
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:35]: general note but add return type for methods

File: projects/fusion-e2e/tests/components/alert/alert.pw.ts [hunk 10]
```diff
@@ -1,4 +1,4 @@
-import {AlertPage} from './alert-page';
+import {AlertPage} from '../../pages/alert-page';
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert.pw.ts:1]: Why is this file here? it looks like a test file - if so it should be located under a relevant folder since this is not something you would like to expose in your library

File: projects/fusion-e2e/tests/pages/alert-page.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            testId: defaultTestId,
+            componentId: alertStoryId,
+            loadedPageSelector: loadedPageSelector
+        };
+
+        super(alertProps);
+        this.component = new AlertComponent(page, this.testId);
+    }
+
+    getAlertText() {
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/pages/alert-page.ts:22]: No need to wrap the same functions if you inherit them from the component class

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 6 above)
Comment (NoaMartan) [inline projects/fusion-e2e/tests/components/alert/alert-component.ts:8]: Important note - it can be very problematic if the classes properties are private, since I want to be able to access the properties and execute methods that exists in baseElement or Button class directly (assuming you didn’t create a method for the same logic of course)

File: projects/fusion-e2e/tests/behavior/behavior-class/base-element.ts [hunk 12]
```diff
@@ -0,0 +1,193 @@
+import {Locator, Page, test} from '@playwright/test';
```
Comment (NoaMartan) [inline projects/fusion-e2e/tests/behavior/behavior-class/base-element.ts:1]: I have a few  notes regarding the consts file . (its not a part of the pr for some reason but it should be)

1. It appears that you don’t use most of the constants there within the alert component so remove them
2. If you do need some of them -  We need to decide if its right to expose the contestants  file to external users - maybe its something that should not be exported as a part of the library ? Not sure lets discuss it
3. Change the name to constants instead of consts :)

**Review by NoaMartan — CHANGES_REQUESTED** [review]

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 5 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts]: Done

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 9 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts]: Done for all methods

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 6 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts]: This is based on the code I took from you, I assumed there was a reason you made it private, but I can remove it and it will be public by default

Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts]: Done for all project

File: projects/fusion-e2e/tests/behavior/behavior-class/base-element.ts (hunk 12 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/behavior/behavior-class/base-element.ts:1]: 1 and 3 Done
2 - lets discuss it

File: projects/fusion-e2e/tests/pages/alert-page.ts (hunk 11 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/pages/alert-page.ts:22]: it's on the page, so we need to use the component class

File: projects/fusion-e2e/tests/components/alert/alert.pw.ts (hunk 10 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert.pw.ts:1]: You're the one who asked me to move the "pages" out yesterday, aren't you?

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 8 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts:28]: I think already done, but lets discuss it

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 6 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts]: It is OK also like this

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 8 above)
Comment (NoaMartan) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts:28]: Unify button and ButtonComponent class

File: projects/fusion-e2e/tests/components/alert/alert.pw.ts (hunk 10 above)
Comment (NoaMartan) [reply projects/fusion-e2e/tests/components/alert/alert.pw.ts:1]: Separate test to its own folder + change tests file name to a indicative name

File: projects/fusion-e2e/tests/pages/alert-page.ts (hunk 11 above)
Comment (NoaMartan) [reply projects/fusion-e2e/tests/pages/alert-page.ts:22]: buttonPage.button.click

change component property name to button

File: projects/fusion-e2e/tests/components/alert/alert-component.ts (hunk 8 above)
Comment (bar-unity) [reply projects/fusion-e2e/tests/components/alert/alert-component.ts:28]: Done, also move all components from demand_platform_ui and fixed dependencies

**Review by AndyKIron — APPROVED** [review]
Will better review who also write tests

**Review by AndyKIron — COMMENTED** [review]
Just not so understand why all under “entities” folder…
